torchbenchmark/models/pytorch_struct/__init__.py

```diff
@@ -24,105 +25,99 @@
+        text_list.append(processed_text)
+        label_list.append(torch.tensor(vocab(selected_labels), dtype=torch.int64))
+    label_list = pad_sequence(label_list, padding_value=padding_value, batch_first=True)
+    text_list = pad_sequence(text_list, padding_value=padding_value, batch_first=True)
+    return text_list.to(device), (text_list != padding_value).sum(axis=1).to(device)
 
 class Model(BenchmarkModel):
-  task = OTHER.OTHER_TASKS
-
-  # Original train batch size: 200
-  # Source: https://github.com/harvardnlp/pytorch-struct/blob/f4e374e894b94a9411fb3d2dfb44201a18e37b26/notebooks/Unsupervised_CFG.ipynb
-  def __init__(self, device=None, jit=False, train_bs=200):
-    super().__init__()
-    self.device = device
-    self.jit = jit
-
-    WORD = Field(include_lengths=True)
-    UD_TAG = Field(init_token="<bos>", eos_token="<eos>", include_lengths=True)
-
-    train, val, test = UDPOS.splits(
-      fields=(('word', WORD), ('udtag', UD_TAG), (None, None)), 
-      filter_pred=lambda ex: 5 < len(ex.word) < 30
-    )
-
-    WORD.build_vocab(train.word, min_freq=3)
-    UD_TAG.build_vocab(train.udtag)
-    self.train_iter = TokenBucket(train, 
-                                  batch_size=train_bs,
-                                  device=self.device)
-
-    # Build model
-    H = 256
-    T = 30
-    NT = 30
-    self.model = NeuralCFG(len(WORD.vocab), T, NT, H)
-    self.model.to(device=device)
-    self.opt = torch.optim.Adam(self.model.parameters(), lr=0.001, betas=[0.75, 0.999])
-    self.train_data = _prefetch(self.train_iter, self.device)
-
-  def get_module(self):
-    for words, _ in self.train_data:
-      return self.model, (words, )
-
-  def train(self, niter=1):
-    if self.jit:
-        raise NotImplementedError("JIT is not supported by this model")
-    for _, (words, lengths) in zip(range(niter), self.train_data):
-      losses = []
-      self.opt.zero_grad()
-      params = self.model(words)
-      dist = SentCFG(params, lengths=lengths)
-      loss = dist.partition.mean()
-      (-loss).backward()
-      losses.append(loss.detach())
-      torch.nn.utils.clip_grad_norm_(self.model.parameters(), 3.0)
-      self.opt.step()
-
-  def eval(self, niter=1):
-    raise NotImplementedError("Eval is not supported by this model")
+    task = OTHER.OTHER_TASKS
+
+    # Original train batch size: 200
+    # Source: https://github.com/harvardnlp/pytorch-struct/blob/f4e374e894b94a9411fb3d2dfb44201a18e37b26/notebooks/Unsupervised_CFG.ipynb
+    def __init__(self, device=None, jit=False, train_bs=40):
+        super().__init__()
+        self.device = device
+        self.jit = jit
+
+        train_iter = UDPOS(split='train')
+        vocab = build_vocab_from_iterator(
+            yield_tokens(train_iter), specials=["<unk>", "<bos>", "<eos>"], min_freq=3
+        )
+        vocab.set_default_index(vocab["<unk>"])
+
+        self.train_data = DataLoader(
+          UDPOS(split='train'),
```

Comment:
Can we have a variable, `prefetch`, to control whether we prefetch the data to GPU in `train()`? Since in TorchBench we only run 1 step (1 epoch and 1 batch), if we don't prefetch data, most of the benchmarking time will be spent on dataloading, making it less sensitive to performance regressions/optimizations for the computation code. So we want to have a variable to control whether we want to prefetch the data to device.

Reply:
In practice, bottleneck on dataloading usually only happens on the first few steps and in later steps it will bottleneck on computation. We want to have a variable to control whether prefetch the data to device to "simulate" a step in the middle of training/inference process.